packages/react-await/src/AwaitClient.ts

```diff
@@ -0,0 +1,123 @@
+import type { AwaitOptions, Key } from './types'
+import { hashKey } from './utils'
+
+type Sync = (...args: unknown[]) => unknown
+
+type AwaitState<TKey extends Key = Key> = {
+  promise?: Promise<unknown>
+  key: TKey
+  hashedKey: ReturnType<typeof hashKey>
+  error?: unknown
+  data?: unknown
+}
+
+class AwaitClient {
+  private cache = new Map<ReturnType<typeof hashKey>, AwaitState>()
+  private syncsMap = new Map<ReturnType<typeof hashKey>, Sync[]>()
+
+  public reset = (key?: Key) => {
+    if (key === undefined || key.length === 0) {
+      this.cache.clear()
+      this.syncSubscribers()
+      return
+    }
+
+    const hashedKey = hashKey(key)
+
+    if (this.cache.has(hashedKey)) {
+      // TODO: reset with key index hierarchy
+      this.cache.delete(hashedKey)
+    }
+
+    this.syncSubscribers(key)
+  }
+
+  public clearError = (key?: Key) => {
+    if (key === undefined || key.length === 0) {
+      this.cache.forEach((value, key, map) => {
+        map.set(key, { ...value, promise: undefined, error: undefined })
+      })
+      return
+    }
+
+    const hashedKey = hashKey(key)
+    const awaitState = this.cache.get(hashedKey)
+    if (awaitState) {
+      // TODO: clearError with key index hierarchy
+      this.cache.set(hashedKey, { ...awaitState, promise: undefined, error: undefined })
+    }
+  }
+
+  public suspend = <TData, TKey extends Key = Key>({ key, fn }: AwaitOptions<TData, TKey>): TData => {
+    const hashedKey = hashKey(key)
+    const awaitState = this.cache.get(hashedKey)
+
+    if (awaitState?.error) {
+      // eslint-disable-next-line @typescript-eslint/no-throw-literal
+      throw awaitState.error
+    }
+    if (awaitState?.data) {
+      return awaitState.data as TData
+    }
+
+    if (awaitState?.promise) {
+      // eslint-disable-next-line @typescript-eslint/no-throw-literal
+      throw awaitState.promise
+    }
+
+    const newAwaitState: AwaitState<TKey> = {
+      key,
+      hashedKey,
+      promise: fn({ key })
+        .then((data) => {
+          newAwaitState.data = data
+        })
+        .catch((error: unknown) => {
+          newAwaitState.error = error
+        }),
+    }
+
+    this.cache.set(hashedKey, newAwaitState)
+    // eslint-disable-next-line @typescript-eslint/no-throw-literal
+    throw newAwaitState.promise
```

Comment:
```suggestion
  public suspend = <TData, TKey extends Key = Key>({ key, fn }: AwaitOptions<TData, TKey>): TData => {
    const hashedKey = hashKey(key)
    const awaitState = this.cache.get(hashedKey)
    
    if(awaitState) { // Ensure awaitState
      if (awaitState.error) {
        // eslint-disable-next-line @typescript-eslint/no-throw-literal
        throw awaitState.error
      }
      if (awaitState.data) {
        return awaitState.data as TData
      }
  
      if (awaitState.promise) {
        // eslint-disable-next-line @typescript-eslint/no-throw-literal
        throw awaitState.promise
      }
    }
    
    const newAwaitState: AwaitState<TKey> = {
      key,
      hashedKey,
      promise: fn({ key })
        .then((data) => {
          newAwaitState.data = data
        })
        .catch((error: unknown) => {
          newAwaitState.error = error
        }),
    }

    this.cache.set(hashedKey, newAwaitState)
    // eslint-disable-next-line @typescript-eslint/no-throw-literal
    throw newAwaitState.promise
```

My small opinion is that ensuring awaitState would be more helpful for debugging when distinguishing between when newAwaitState is used and when awaitState is used. What do you think?